subprojects/base-services/src/main/java/org/gradle/internal/hash/Hasher.java

```diff
@@ -21,52 +21,7 @@
  *
  * In order to avoid collisions we prepend the length of the next bytes to the underlying hasher (see this <a href="http://crypto.stackexchange.com/a/10065">answer</a> on stackexchange).
  */
-public interface Hasher {
-    /**
-     * Feed a bunch of bytes into the hasher.
-     */
-    void putBytes(byte[] bytes);
-
-    /**
-     * Feed a given number of bytes into the hasher from the given offset.
-     */
-    void putBytes(byte[] bytes, int off, int len);
-
-    /**
-     * Feed a single byte into the hasher.
-     */
-    void putByte(byte value);
-
-    /**
-     * Feed an integer byte into the hasher.
-     */
-    void putInt(int value);
-
-    /**
-     * Feed a long value byte into the hasher.
-     */
-    void putLong(long value);
-
-    /**
-     * Feed a double value into the hasher.
-     */
-    void putDouble(double value);
-
-    /**
-     * Feed a boolean value into the hasher.
-     */
-    void putBoolean(boolean value);
-
-    /**
-     * Feed a string into the hasher.
-     */
-    void putString(CharSequence value);
-
-    /**
-     * Feed a hash code into the hasher.
-     */
-    void putHash(HashCode hashCode);
-
+public interface Hasher extends PrimitiveHasher {
```

Comment:
This violates the LSP (https://en.wikipedia.org/wiki/Liskov_substitution_principle): A Hasher is not a PrimitiveHasher, it behaves differently.

Reply:
I understood that `Hasher` does hashing of primitives and does more (nulls and errors), but can function as a primitive hasher as well.
Do you think I should rename `PrimitiveHasher`, or revert the extends?

Reply:
Hasher and PrimitiveHasher have different contracts. The former guarantees that it delimits entries, the latter guarantees that it doesn't. One can't extend the other.

Reply:
You're right, thanks for bringing it to my attention.
Removed the benchmarks, too.